### PR TITLE
transaction: fix `TxOut::is_pegout()` to fail properly on certain invalid inputs

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

File: src/transaction.rs [hunk 1]
```diff
@@ -1440,5 +1440,55 @@ mod tests {
...
+            edededededededede9edededede00000000000001000000000000050000ff0000\
+            000000000000000000ff000000000000000000000000200000000000011c00000\
+            000d3d3d3d3d3d3d3d3d3d3d3d3d3d3d3d3d3d3f3d3d3d3d3d3d3d3d3d3d3d3d3\
+            d3d3d3d3d3d3\
+        ");
+
+        assert!(!output.is_null_data());
+        assert!(!output.is_pegout());
+
+        // Output with pushes that are e.g. OP_1 are nulldata but not pegouts
```
Comment (jonasnick) [inline src/transaction.rs]: I don't see an OP_1 in the hex.

File: src/transaction.rs [hunk 2]
```diff
@@ -353,7 +353,7 @@ impl TxOut {
         if iter.next() == Some(Instruction::Op(opcodes::All::OP_RETURN)) {
             for push in iter {
                 match push {
-                    Instruction::Op(op) if op as u8 > opcodes::All::OP_PUSHNUM_16 as u8 => {}
+                    Instruction::Op(op) if op as u8 > opcodes::All::OP_PUSHNUM_16 as u8 => return false,
```
Comment (jonasnick) [inline src/transaction.rs]: Can we add an `Instruction::Op(op) if op as u8 <= opcodes::All::OP_PUSHNUM_16 as u8 => return true` case. I find it hard to see what isan Op or PushBytes. I think OP_1 aka OP_PUSHNUM_1 is not PushBytes and therefore would result in false here.

Comment (apoelstra) [reply src/transaction.rs]: Opcodes 0 through 76 are "push this many bytes", followed by the 3 PUSHDATA opcodes, a couple reserved ones, then the numbers. Everything above OP_16 is control flow and OP_RETURN. OP_PUSHNUM_1 0x51 is definitely belowe OP_PUSHNUM_16 0x60.

I copied this logic from https://github.com/bitcoin/bitcoin/blob/master/src/script/script.cpp#L243 btw.

File: src/transaction.rs (hunk 1 above)
Comment (apoelstra) [reply src/transaction.rs]: Oh, oops, it's actually pushing a literal 0x01. Good call, this is not what I meant to test.

File: src/transaction.rs (hunk 2 above)
Comment (apoelstra) [reply src/transaction.rs]: Oh, I understand what you're saying! Will fix.

**Comment by apoelstra** [comment]
Fixed @jonasnick comments. The logic should actually be correct now :)

**Comment by jonasnick** [comment]
ACK

**Pull request merged**